File: 백준/Gold/22866. 탑 보기/탑 보기.java
import java.util.*;
import java.io.*;

public class Main {
    static int n = 0;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        int[] h = new int[n+1];
        int[] num = new int[n+1];
        int[][] s = new int[n+1][2];

        for(int i=1;i<=n;i++)
        {
            s[i][1] = Integer.MAX_VALUE;
        }

        st = new StringTokenizer(br.readLine(), " ");

        for(int i=1;i<=n;i++)
        {
            h[i] = Integer.parseInt(st.nextToken());
        }

        Stack<Point> stack = new Stack<>();

        for(int i=1;i<=n;i++)
        {
            while(!stack.isEmpty())
            {
                Point cur = stack.peek();

                if(cur.h <= h[i])
                {
                    stack.pop();
                }

                else{
                    if(Math.abs(cur.idx - i) < s[i][1])
                    {
                        s[i][0] = cur.idx;
                        s[i][1] = Math.abs(cur.idx - i);
                    }
                    num[i] += stack.size();
                    break;
                }

            }

            stack.add(new Point(i,h[i]));
        }

        stack.clear();

        for(int i=n;i>=1;i--)
        {
            while(!stack.isEmpty())
            {
                Point cur = stack.peek();

                if(cur.h <= h[i])
                {
                    stack.pop();
                }

                else{
                    if(Math.abs(cur.idx - i) < s[i][1])
                    {
                        s[i][0] = cur.idx;
                        s[i][1] = Math.abs(cur.idx - i);
                    }

                    else if(Math.abs(cur.idx - i) == s[i][1])
                    {
                        s[i][0] = Math.min(cur.idx , s[i][0]);
                    }
                    num[i] += stack.size();
                    break;
                }
            }

            stack.add(new Point(i,h[i]));
        }

        StringBuilder sb = new StringBuilder();
        for(int i=1;i<=n;i++) {
           if(num[i] == 0)
                sb.append(0+"\n");

           else
               sb.append(num[i]+" "+s[i][0]+"\n");

        }
        System.out.print(sb);
        br.close();
    }
}
class Point
{
    int idx;
    int h;

    Point(int idx,int h)
    {
        this.idx = idx;
        this.h = h;
    }
}

File: 백준/Gold/22866. 탑 보기/탑 보기.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

data class Point(val idx: Int, val h: Int)

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine(), " ")
    val n = st.nextToken().toInt()
    val h = IntArray(n + 1)
    val num = IntArray(n + 1)
    val s = Array(n + 1) { IntArray(2) { Int.MAX_VALUE } }

    st = StringTokenizer(br.readLine(), " ")
    for (i in 1..n) {
        h[i] = st.nextToken().toInt()
    }

    val stack = Stack<Point>()

    for (i in 1..n) {
        while (stack.isNotEmpty()) {
            val cur = stack.peek()
            if (cur.h <= h[i]) {
                stack.pop()
            } else {
                if (Math.abs(cur.idx - i) < s[i][1]) {
                    s[i][0] = cur.idx
                    s[i][1] = Math.abs(cur.idx - i)
                }
                num[i] += stack.size
                break
            }
        }
        stack.add(Point(i, h[i]))
    }

    stack.clear()

    for (i in n downTo 1) {
        while (stack.isNotEmpty()) {
            val cur = stack.peek()
            if (cur.h <= h[i]) {
                stack.pop()
            } else {
                if (Math.abs(cur.idx - i) < s[i][1]) {
                    s[i][0] = cur.idx
                    s[i][1] = Math.abs(cur.idx - i)
                } else if (Math.abs(cur.idx - i) == s[i][1]) {
                    s[i][0] = Math.min(cur.idx, s[i][0])
                }
                num[i] += stack.size
                break
            }
        }
        stack.add(Point(i, h[i]))
    }

    val sb = StringBuilder()
    for (i in 1..n) {
        if (num[i] == 0) {
            sb.append("0\n")
        } else {
            sb.append("${num[i]} ${s[i][0]}\n")
        }
    }
    print(sb)
    br.close()
}
